### PR TITLE
expression: Support GBK for builtin function MD5

### LINKED ISSUE #29090: Support GBK for builtin function `MD5`

### PR DESCRIPTION
**Pull request by Defined2014**
### What problem does this PR solve?

Issue Number: close #29090

Problem Summary: Support GBK for builtin function MD5

### Check List

Tests 

- [x] Unit test
- [x] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No code

### Release note

```release-note
None
```

### CODE REVIEW

File: expression/builtin_encryption.go [hunk 1]
```diff
@@ -625,7 +626,13 @@ func (b *builtinMD5Sig) evalString(row chunk.Row) (string, bool, error) {
 	if isNull || err != nil {
 		return "", isNull, err
 	}
-	sum := md5.Sum([]byte(arg)) // #nosec G401
+	var sum [16]byte
+	dBytes, err := charset.NewEncoding(b.args[0].GetType().Charset).Encode(nil, []byte(arg))
+	if err == nil {
+		sum = md5.Sum(dBytes) // #nosec G401
+	} else {
```
Comment (zimulala) [inline expression/builtin_encryption.go:633]: Do we have a test for this logic?

Comment (Defined2014) [reply expression/builtin_encryption.go:633]: No, we will not go to this branch in TiDB normal use, because of #28778. If it's necessary, I will add it.

Comment (Defined2014) [reply expression/builtin_encryption.go:633]: And a question here, need I add a warning log when get invalid character here?

Comment (zimulala) [reply expression/builtin_encryption.go:633]: @hawkingrei https://github.com/pingcap/tidb/issues/29314 will handle this problem uniformly? Maybe we can return an error or a panic？

Comment (Defined2014) [reply expression/builtin_encryption.go]: @zimulala Updated PTAL

Comment (hawkingrei) [reply expression/builtin_encryption.go]: > @hawkingrei #29314 will handle this problem uniformly? Maybe we can return an error or a panic？

OK, I will.

**Review by zimulala — APPROVED** [review]
LGTM

**Review by tangenta — APPROVED** [review]

**Pull request merged**